src/libs/protoboard/blackboard_manager.h

```diff
@@ -0,0 +1,331 @@
+#include <aspect/configurable.h>
+#include <aspect/logging.h>
+#include <blackboard/utils/on_message_waker.h>
+#include <core/threading/thread.h>
+#include <interfaces/ProtobufPeerInterface.h>
+
+#include <boost/fusion/include/any.hpp>
+#include <boost/fusion/include/for_each.hpp>
+#include <boost/fusion/include/std_tuple.hpp>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+namespace protoboard {
+
+/** Map a blackboard interface type to a blackboard interface ID.
+ * Must be implemented by the user. Will be called for every type used with
+ * an @a bb_iface_manager.
+ * @return The interface name (ID) that should be used for the given @tparam type */
+template <class IfaceT>
+std::string iface_id_for_type();
+
+/** Must be implemented by the user.
+ * @return a vector of paths where ProtoBuf should look for message definitions */
+std::vector<std::string> proto_dirs();
+
+/**
+ * Container for an opened interface of type @tparam IfaceT.
+ * @tparam MessageTypeList must be a @a type_list of the message types that should be
+ * handled on @tparam IfaceT.
+ */
+template <class IfaceT, class MessageTypeList>
+class bb_iface_manager
+{
+public:
+	/// Constructor. Not responsible for actual initialization.
+	bb_iface_manager() : interface_(nullptr), blackboard_(nullptr), waker_(nullptr)
+	{
+	}
+
+	/** Open an interface of the given type with the ID supplied by @a iface_id_for_type and
+	 * register to wake the given thread when any of the given types arrives.
+	 * @param blackboard The blackboard to use
+	 * @param thread The thread to wake */
+	void
+	init(fawkes::BlackBoard *blackboard, fawkes::Thread *thread)
+	{
+		blackboard_ = blackboard;
+		interface_  = blackboard_->open_for_writing<IfaceT>(iface_id_for_type<IfaceT>().c_str());
+		waker_      = new fawkes::BlackBoardOnMessageWaker(blackboard, interface_, thread);
+	}
+
+	/// Cleanup.
+	void
+	finalize()
+	{
+		if (waker_)
+			delete waker_;
+		if (blackboard_ && interface_)
+			blackboard_->close(interface_);
+	}
+
+	/// @return The managed interface
+	IfaceT *
+	interface() const
+	{
+		return interface_;
+	}
+
+private:
+	IfaceT *                          interface_;
+	fawkes::BlackBoard *              blackboard_;
+	fawkes::BlackBoardOnMessageWaker *waker_;
+};
+
+/**
+ * Abstract superclass for sending out ProtoBuf messages
+ */
+class AbstractProtobufSender
+{
+public:
+	/** Constructor.
+	 * @param bb_mgr The BlackboardManager that uses this */
+	AbstractProtobufSender(BlackboardManager *bb_mgr);
+
+	/// Destructor
+	virtual ~AbstractProtobufSender();
+
+	/** Go through all interface managers, empty all blackboard message queues and send out
+	 * ProtoBuf messages accordingly.
+	 * @return whether anything was sent */
+	virtual bool process_sending_interfaces() = 0;
+
+	/// Deferred initialization, coincides with the main thread.
+	virtual void init() = 0;
+	/// Deferred cleanup, concides with the main thread.
+	virtual void finalize() = 0;
```

Comment:
Similar to the previous remark: It might be possible to get rid of it, but I think that would entail a major internal redesign. Something that can definitely be considered, but not at this point in time.